### PR TITLE
Update to CAPI 1.5

### PR DESCRIPTION
**Pull request by hrak**
*Issue #, if available:*

#285
#253

*Description of changes:*

This PR updates the cluster-api dependency to v1.5.5. It also updates to Go 1.20, updates golang-ci and addresses lint issues caught by the new version (mostly unused variables in function definitions)

*Testing performed:*

make test
test with kind

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by hrak** [comment]
whoa, i seem to have awoken the codecov reporter :)

**Comment by rohityadavcloud** [comment]
I haven't tested it, but in theory we must upgrade to CAPI 1.5 or whatever may be the latest. I would want to see an easily v0.5 release.

/lgtm

**Comment by rohityadavcloud** [comment]
cc @shwstppr we'll need that CI fixed soon 🙏

**Comment by g-gaston** [comment]
/run-e2e -c 4.18
let's see if we get anything

**Comment by pokearu** [comment]
Hi @g-gaston @rohityadavcloud I ran the e2e suit manually to test the CAPI changes. And got a successful run ✅

Not sure why the presubmit is failing, `make run-e2e-smoke` works for me locally just fine.

**Comment by rohityadavcloud** [comment]
Thanks for testing @pokearu, @g-gaston @hrak is this ready for merging or need further review/testing? There was also an idea to skip and move to CAPI 1.6 ?

**Comment by pokearu** [comment]
@rohityadavcloud looks like the failing presubmit is due to the version of KinD being installed.
https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/blob/main/hack/ensure-kind.sh#L22
looks like this needs to be increased to `v0.20.0` to match with CAPI 1.5.5

**Comment by g-gaston** [comment]
@hrak great job!
I added one small commit to fix the presubmits, just bumping the kind version
Would you mind squashing the commits and rebasing?

File: controllers/utils/base_reconciler.go [hunk 1]
```diff
@@ -44,11 +44,11 @@ import (
...
-	BaseLogger logr.Logger
-	Scheme     *runtime.Scheme
-	K8sClient  client.Client
-	CSClient   cloud.Client
-	Recorder   record.EventRecorder
+	Scheme           *runtime.Scheme
+	K8sClient        client.Client
+	CSClient         cloud.Client
+	Recorder         record.EventRecorder
+	WatchFilterValue string
```
Comment (g-gaston) [inline controllers/utils/base_reconciler.go:51]: what is this new field for?
if it's not used by the base reconciler, can we put it where it's used?

File: controllers/utils/utils.go [hunk 2]
```diff
@@ -175,3 +184,67 @@ func WithClusterSuffix(name string, clusterName string) string {
...
+		if gv.Group == clusterv1.GroupVersion.Group {
+			return ref.Name, nil
+		}
+	}
+	return "", errors.New("failed to get owner Cluster name")
+}
+
+// CloudStackClusterToCloudStackMachines is a handler.ToRequestsFunc to be used to enqeue requests for reconciliation
+// of CloudStackMachines.
+func CloudStackClusterToCloudStackMachines(c client.Client, obj runtime.Object, scheme *runtime.Scheme, log logr.Logger) (handler.MapFunc, error) {
```
Comment (g-gaston) [inline controllers/utils/utils.go:207]: can we add unit test for these 2 new methods?

**Comment by hrak** [comment]
PR updated for CAPI v1.5.8

File: controllers/cloudstackmachine_controller.go [hunk 3]
```diff
@@ -348,14 +354,48 @@ func (r *CloudStackMachineReconciliationRunner) ReconcileDelete() (retRes ctrl.R
 
 // SetupWithManager registers the machine reconciler to the CAPI controller manager.
 func (reconciler *CloudStackMachineReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, opts controller.Options) error {
-	log := ctrl.LoggerFrom(ctx)
+	reconciler.Recorder = mgr.GetEventRecorderFor("capc-machine-controller")
+	CloudStackClusterToCloudStackMachines, err := utils.CloudStackClusterToCloudStackMachines(reconciler.K8sClient, &infrav1.CloudStackMachineList{}, reconciler.Scheme, ctrl.LoggerFrom(ctx))
+	if err != nil {
+		return errors.Wrap(err, "failed to create CloudStackClusterToCloudStackMachines mapper")
+	}
+	//requeueCloudStackMachinesForUnpausedCluster := reconciler.requeueCloudStackMachinesForUnpausedCluster(ctx)
```
Comment (vishesh92) [inline controllers/cloudstackmachine_controller.go:362]: Remove commented code.

**Comment by vishesh92** [comment]
@hrak there are some unresolved comments by @g-gaston. Can you please address/resolve them?

**Comment by vishesh92** [comment]
This PR has some conflicts and some comments needs to be addressed/resolved. Changing milestone for this PR to v0.6.0 for now.

**Comment by hrak** [comment]
Hi, I'm sorry I haven't been able to dedicate much time on this lately cause I'm swamped in launching our k8s product at work. I will update the PR when i have some time.

**Pull request closed without merging**